server/src/main/java/org/elasticsearch/cluster/metadata/InferenceFieldMetadata.java

```diff
@@ -0,0 +1,125 @@
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+public final class InferenceFieldMetadata implements SimpleDiffable<InferenceFieldMetadata>, ToXContentFragment {
+    private static final String INFERENCE_ID_FIELD = "inference_id";
+    private static final String SOURCE_FIELDS_FIELD = "source_fields";
+
+    private final String name;
+    private final String inferenceId;
+    private final String[] sourceFields;
+
+    public InferenceFieldMetadata(String name, String inferenceId, String[] sourceFields) {
+        this.name = Objects.requireNonNull(name);
+        this.inferenceId = Objects.requireNonNull(inferenceId);
+        this.sourceFields = Objects.requireNonNull(sourceFields);
+    }
+
+    public InferenceFieldMetadata(StreamInput input) throws IOException {
+        this.name = input.readString();
+        this.inferenceId = input.readString();
+        this.sourceFields = input.readStringArray();
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(name);
+        out.writeString(inferenceId);
+        out.writeStringArray(sourceFields);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        InferenceFieldMetadata that = (InferenceFieldMetadata) o;
+        return inferenceId.equals(that.inferenceId) && Arrays.equals(sourceFields, that.sourceFields);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = Objects.hash(inferenceId);
+        result = 31 * result + Arrays.hashCode(sourceFields);
+        return result;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public String getInferenceId() {
+        return inferenceId;
+    }
+
+    public String[] getSourceFields() {
+        return sourceFields;
+    }
+
+    public static Diff<InferenceFieldMetadata> readDiffFrom(StreamInput in) throws IOException {
+        return SimpleDiffable.readDiffFrom(InferenceFieldMetadata::new, in);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(name);
+        builder.field(INFERENCE_ID_FIELD, inferenceId);
+        builder.array(SOURCE_FIELDS_FIELD, sourceFields);
+        return builder.endObject();
+    }
+
+    public static InferenceFieldMetadata fromXContent(XContentParser parser) throws IOException {
+        final String name = parser.currentName();
+
+        XContentParser.Token token = parser.nextToken();
+        if (token == null) {
+            // no data...
+            return null;
+        }
+        String currentFieldName = null;
+        String inferenceId = null;
+        List<String> inputFields = new ArrayList<>();
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
```

Comment:
Is there some guidance on when to use a `ConstructingObjectParser` vs doing "manual" parsing?

Reply:
ah, got it. Thanks!